File: node/src/chaincodeevents.test.ts
/*
 * Copyright 2021 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { CallOptions, Metadata, ServiceError, status } from '@grpc/grpc-js';
import { gateway as gatewayproto, orderer, peer } from '@hyperledger/fabric-protos';
import { ChaincodeEvent } from './chaincodeevent';
import * as checkpointers from './checkpointers';
import { Gateway, InternalConnectOptions, internalConnect } from './gateway';
import { GatewayError } from './gatewayerror';
import { Identity } from './identity/identity';
import { Network } from './network';
import { MockGatewayGrpcClient, newServerStreamResponse, readElements } from './testutils.test';

function assertDecodeChaincodeEventsRequest(
    signedRequest: gatewayproto.SignedChaincodeEventsRequest,
): gatewayproto.ChaincodeEventsRequest {
    const requestBytes = signedRequest.getRequest_asU8();
    expect(requestBytes).toBeDefined();
    return gatewayproto.ChaincodeEventsRequest.deserializeBinary(requestBytes);
}

function newChaincodeEvent(blockNumber: number, event: peer.ChaincodeEvent): ChaincodeEvent {
    return {
        blockNumber: BigInt(blockNumber),
        chaincodeName: event.getChaincodeId(),
        eventName: event.getEventName(),
        transactionId: event.getTxId(),
        payload: event.getPayload_asU8(),
    };
}

interface ExpectedRequest {
    channelName: string;
    chaincodeName: string;
    typeCase: orderer.SeekPosition.TypeCase;
    blockNumber?: bigint;
    transactionId?: string;
}

function assertChaincodeEventRequest(
    actual: gatewayproto.ChaincodeEventsRequest,
    expectedRequest: ExpectedRequest,
): void {
    expect(actual.getChannelId()).toBe(expectedRequest.channelName);
    expect(actual.getChaincodeId()).toBe(expectedRequest.chaincodeName);

    const startPosition = actual.getStartPosition();

    expect(startPosition).toBeDefined();
    expect(startPosition?.getTypeCase()).toBe(expectedRequest.typeCase);
    if (expectedRequest.blockNumber != undefined) {
        expect(startPosition?.getSpecified()?.getNumber()).toBe(Number(expectedRequest.blockNumber));
    }
    if (expectedRequest.transactionId) {
        expect(actual.getAfterTransactionId()).toEqual(expectedRequest.transactionId);
    }
}

describe('Chaincode Events', () => {
    const channelName = 'CHANNEL_NAME';
    const signature = Buffer.from('SIGNATURE');
    const serviceError: ServiceError = Object.assign(new Error('ERROR_MESSAGE'), {
        code: status.UNAVAILABLE,
        details: 'DETAILS',
        metadata: new Metadata(),
    });

    let chaincodeEventsOptions: () => CallOptions;
    let client: MockGatewayGrpcClient;
    let identity: Identity;
    let gateway: Gateway;
    let network: Network;

    const event1 = new peer.ChaincodeEvent();
    event1.setChaincodeId('CHAINCODE');
    event1.setTxId('tx1');
    event1.setEventName('event1');
    event1.setPayload(Buffer.from('payload1'));

    const event2 = new peer.ChaincodeEvent();
    event2.setChaincodeId('CHAINCODE');
    event2.setTxId('tx2');
    event2.setEventName('event2');
    event2.setPayload(Buffer.from('payload2'));

    const event3 = new peer.ChaincodeEvent();
    event3.setChaincodeId('CHAINCODE');
    event3.setTxId('tx3');
    event3.setEventName('event3');
    event3.setPayload(Buffer.from('payload3'));

    beforeEach(() => {
        const now = new Date();
        const callOptions = {
            deadline: now.setHours(now.getHours() + 1),
        };
        chaincodeEventsOptions = () => callOptions; // Return a specific object to test modification

        client = new MockGatewayGrpcClient();
        identity = {
            mspId: 'MSP_ID',
            credentials: Buffer.from('CERTIFICATE'),
        };

        const options: InternalConnectOptions = {
            identity,
            signer: () => Promise.resolve(signature),
            client,
            chaincodeEventsOptions,
        };
        gateway = internalConnect(options);
        network = gateway.getNetwork(channelName);
    });

    describe('request', () => {
        it('sends valid request with default start position', async () => {
            await network.getChaincodeEvents('CHAINCODE');

            const signedRequest = client.getChaincodeEventsRequests()[0];
            expect(signedRequest.getSignature()).toEqual(signature);

            const expected: ExpectedRequest = {
                channelName: channelName,
                chaincodeName: 'CHAINCODE',
                typeCase: orderer.SeekPosition.TypeCase.NEXT_COMMIT,
            };

            const request = assertDecodeChaincodeEventsRequest(signedRequest);
            assertChaincodeEventRequest(request, expected);
        });

        it('throws with negative specified start block number', async () => {
            const startBlock = BigInt(-1);
            return expect(network.getChaincodeEvents('CHAINCODE', { startBlock })).rejects.toThrow();
        });

        it('sends valid request with specified start block number', async () => {
            const startBlock = BigInt(418);

            await network.getChaincodeEvents('CHAINCODE', { startBlock });

            const signedRequest = client.getChaincodeEventsRequests()[0];
            expect(signedRequest.getSignature()).toEqual(signature);

            const expected: ExpectedRequest = {
                channelName: channelName,
                chaincodeName: 'CHAINCODE',
                typeCase: orderer.SeekPosition.TypeCase.SPECIFIED,
                blockNumber: startBlock,
            };

            const request = assertDecodeChaincodeEventsRequest(signedRequest);
            assertChaincodeEventRequest(request, expected);
        });

        it('Sends valid request with specified start block number and fresh checkpointer', async () => {
            const startBlock = BigInt(418);
            const checkpointer = checkpointers.inMemory();

            await network.getChaincodeEvents('CHAINCODE', { startBlock: startBlock, checkpoint: checkpointer });

            const signedRequest = client.getChaincodeEventsRequests()[0];
            expect(signedRequest.getSignature()).toEqual(signature);

            const expected: ExpectedRequest = {
                channelName: channelName,
                chaincodeName: 'CHAINCODE',
                typeCase: orderer.SeekPosition.TypeCase.SPECIFIED,
                blockNumber: startBlock,
                transactionId: '',
            };
            const request = assertDecodeChaincodeEventsRequest(signedRequest);
            assertChaincodeEventRequest(request, expected);
        });

        it('Sends valid request with specified start block and checkpointed block', async () => {
            const startBlock = BigInt(418);
            const checkpointer = checkpointers.inMemory();
            await checkpointer.checkpointBlock(1n);

            await network.getChaincodeEvents('CHAINCODE', { startBlock: startBlock, checkpoint: checkpointer });

            const signedRequest = client.getChaincodeEventsRequests()[0];
            expect(signedRequest.getSignature()).toEqual(signature);

            const expected: ExpectedRequest = {
                channelName: channelName,
                chaincodeName: 'CHAINCODE',
                typeCase: orderer.SeekPosition.TypeCase.SPECIFIED,
                blockNumber: 1n + 1n,
                transactionId: '',
            };
            const request = assertDecodeChaincodeEventsRequest(signedRequest);
            assertChaincodeEventRequest(request, expected);
        });

        it('Sends valid request with specified start block and checkpointed transaction id', async () => {
            const startBlock = BigInt(418);
            const checkpointer = checkpointers.inMemory();
            await checkpointer.checkpointTransaction(1n, 'txn1');

            await network.getChaincodeEvents('CHAINCODE', { startBlock: startBlock, checkpoint: checkpointer });

            const signedRequest = client.getChaincodeEventsRequests()[0];
            expect(signedRequest.getSignature()).toEqual(signature);

            const expected: ExpectedRequest = {
                channelName: channelName,
                chaincodeName: 'CHAINCODE',
                typeCase: orderer.SeekPosition.TypeCase.SPECIFIED,
                blockNumber: 1n,
                transactionId: 'txn1',
            };
            const request = assertDecodeChaincodeEventsRequest(signedRequest);
            assertChaincodeEventRequest(request, expected);
        });

        it('Sends valid request with no start block and fresh checkpointer', async () => {
            const checkpointer = checkpointers.inMemory();

            await network.getChaincodeEvents('CHAINCODE', { checkpoint: checkpointer });

            const signedRequest = client.getChaincodeEventsRequests()[0];
            expect(signedRequest.getSignature()).toEqual(signature);

            const expected: ExpectedRequest = {
                channelName: channelName,
                chaincodeName: 'CHAINCODE',
                typeCase: orderer.SeekPosition.TypeCase.NEXT_COMMIT,
            };

            const request = assertDecodeChaincodeEventsRequest(signedRequest);
            assertChaincodeEventRequest(request, expected);
        });

        it('Sends valid request with no start block and checkpointer transaction ID', async () => {
            const checkpointer = checkpointers.inMemory();
            await checkpointer.checkpointTransaction(1n, 'txn1');

            await network.getChaincodeEvents('CHAINCODE', { checkpoint: checkpointer });

            const signedRequest = client.getChaincodeEventsRequests()[0];
            expect(signedRequest.getSignature()).toEqual(signature);

            const expected: ExpectedRequest = {
                channelName: channelName,
                chaincodeName: 'CHAINCODE',
                typeCase: orderer.SeekPosition.TypeCase.SPECIFIED,
                blockNumber: 1n,
                transactionId: 'txn1',
            };

            const request = assertDecodeChaincodeEventsRequest(signedRequest);
            assertChaincodeEventRequest(request, expected);
        });

        it('Sends valid request with with start block and checkpointer chaincode event', async () => {
            const checkpointer = checkpointers.inMemory();
            const event: ChaincodeEvent = {
                blockNumber: BigInt(1),
                chaincodeName: 'chaincode',
                eventName: 'event1',
                transactionId: 'txn1',
                payload: new Uint8Array(),
            };

            await checkpointer.checkpointChaincodeEvent(event);
            await network.getChaincodeEvents('CHAINCODE', { checkpoint: checkpointer });

            const signedRequest = client.getChaincodeEventsRequests()[0];
            expect(signedRequest.getSignature()).toEqual(signature);

            const expected: ExpectedRequest = {
                channelName: channelName,
                chaincodeName: 'CHAINCODE',
                typeCase: orderer.SeekPosition.TypeCase.SPECIFIED,
                blockNumber: 1n,
                transactionId: 'txn1',
            };

            const request = assertDecodeChaincodeEventsRequest(signedRequest);
            assertChaincodeEventRequest(request, expected);
        });

        it('uses specified call options', async () => {
            const deadline = Date.now() + 1000;

            await network.newChaincodeEventsRequest('CHAINCODE').getEvents({ deadline });

            const actual = client.getChaincodeEventsOptions()[0];
            expect(actual.deadline).toBe(deadline);
        });

        it('uses default call options', async () => {
            await network.getChaincodeEvents('CHAINCODE');

            const actual = client.getChaincodeEventsOptions()[0];
            expect(actual.deadline).toBe(chaincodeEventsOptions().deadline);
        });

        it('default call options are not modified', async () => {
            const expected = chaincodeEventsOptions().deadline;
            const deadline = Date.now() + 1000;

            await network.newChaincodeEventsRequest('CHAINCODE').getEvents({ deadline });

            expect(chaincodeEventsOptions().deadline).toBe(expected);
        });

        it('throws GatewayError on call ServiceError', async () => {
            client.mockChaincodeEventsError(serviceError);

            const t = network.getChaincodeEvents('CHAINCODE');

            await expect(t).rejects.toThrow(GatewayError);
            await expect(t).rejects.toThrow(serviceError.message);
            await expect(t).rejects.toMatchObject({
                code: serviceError.code,
                cause: serviceError,
            });
        });
    });

    describe('event delivery', () => {
        const response1 = new gatewayproto.ChaincodeEventsResponse();
        response1.setBlockNumber(1);
        response1.setEventsList([event1, event2]);

        const response2 = new gatewayproto.ChaincodeEventsResponse();
        response2.setBlockNumber(2);
        response2.setEventsList([event3]);

        const expectedEvents = [
            newChaincodeEvent(1, event1),
            newChaincodeEvent(1, event2),
            newChaincodeEvent(2, event3),
        ];

        it('returns events as AsyncIterable', async () => {
            client.mockChaincodeEventsResponse(newServerStreamResponse([response1, response2]));

            const events = await network.getChaincodeEvents('CHAINCODE');

            const actualEvents = await readElements(events, expectedEvents.length);
            expect(actualEvents).toEqual(expectedEvents);
        });

        it('closing iterable cancels gRPC client stream', async () => {
            const responseStream = newServerStreamResponse([response1, response2]);
            client.mockChaincodeEventsResponse(responseStream);

            const events = await network.getChaincodeEvents('CHAINCODE');
            events.close();

            expect(responseStream.cancel).toHaveBeenCalled();
        });

        it('throws GatewayError on call ServiceError', async () => {
            client.mockChaincodeEventsResponse(
                newServerStreamResponse<gatewayproto.ChaincodeEventsResponse>([serviceError]),
            );

            const events = await network.getChaincodeEvents('CHAINCODE');
            const t = readElements(events, 1);

            await expect(t).rejects.toThrow(GatewayError);
            await expect(t).rejects.toThrow(serviceError.message);
            await expect(t).rejects.toMatchObject({
                code: serviceError.code,
                cause: serviceError,
            });
        });

        it('resource clean-up cancels gRPC client stream', async () => {
            const responseStream = newServerStreamResponse([response1, response2]);
            client.mockChaincodeEventsResponse(responseStream);

            {
                // @ts-expect-error Assigned to unused variable for resource cleanup
                // eslint-disable-next-line @typescript-eslint/no-unused-vars
                using events = await network.getChaincodeEvents('CHAINCODE');
            }

            expect(responseStream.cancel).toHaveBeenCalled();
        });
    });
});

File: pkg/client/chaincodeevents_test.go
// Copyright IBM Corp. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package client

import (
	"context"
	"errors"
	"io"
	"testing"

	"github.com/hyperledger/fabric-protos-go-apiv2/gateway"
	"github.com/hyperledger/fabric-protos-go-apiv2/orderer"
	"github.com/hyperledger/fabric-protos-go-apiv2/peer"
	"github.com/stretchr/testify/require"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/proto"
)

func TestChaincodeEvents(t *testing.T) {
	newChaincodeEventsResponse := func(events []*ChaincodeEvent) *gateway.ChaincodeEventsResponse {
		blockNumber := uint64(0)
		var peerEvents []*peer.ChaincodeEvent

		for _, event := range events {
			blockNumber = event.BlockNumber
			peerEvents = append(peerEvents, &peer.ChaincodeEvent{
				ChaincodeId: event.ChaincodeName,
				TxId:        event.TransactionID,
				EventName:   event.EventName,
				Payload:     event.Payload,
			})
		}

		return &gateway.ChaincodeEventsResponse{
			BlockNumber: blockNumber,
			Events:      peerEvents,
		}
	}

	t.Run("Returns connect error", func(t *testing.T) {
		expected := NewStatusError(t, codes.Aborted, "CHAINCODE_EVENTS_ERROR")

		mockConnection := NewMockClientConnInterface(t)
		ExpectChaincodeEvents(mockConnection, WithNewStreamError(expected))

		ctx, cancel := context.WithCancel(context.Background())
		defer cancel()

		network := AssertNewTestNetwork(t, "NETWORK", WithClientConnection(mockConnection))
		_, err := network.ChaincodeEvents(ctx, "CHAINCODE")

		require.Equal(t, status.Code(expected), status.Code(err), "status code")
		require.ErrorIs(t, err, expected, "error type: %T", err)
		require.ErrorContains(t, err, expected.Error(), "message")
	})

	for testName, testCase := range map[string]struct {
		options  []ChaincodeEventsOption
		expected *gateway.ChaincodeEventsRequest
	}{
		"Sends valid request with default start position": {
			options: nil,
			expected: &gateway.ChaincodeEventsRequest{
				StartPosition: &orderer.SeekPosition{
					Type: &orderer.SeekPosition_NextCommit{
						NextCommit: &orderer.SeekNextCommit{},
					},
				},
			},
		},
		"Sends valid request with specified start block number": {
			options: []ChaincodeEventsOption{
				WithStartBlock(418),
			},
			expected: &gateway.ChaincodeEventsRequest{
				StartPosition: &orderer.SeekPosition{
					Type: &orderer.SeekPosition_Specified{
						Specified: &orderer.SeekSpecified{
							Number: 418,
						},
					},
				},
			},
		},
		"Sends valid request with specified start block number and fresh checkpointer": {
			options: []ChaincodeEventsOption{
				WithStartBlock(418),
				WithCheckpoint(new(InMemoryCheckpointer)),
			},
			expected: &gateway.ChaincodeEventsRequest{
				StartPosition: &orderer.SeekPosition{
					Type: &orderer.SeekPosition_Specified{
						Specified: &orderer.SeekSpecified{
							Number: 418,
						},
					},
				},
			},
		},
		"Sends valid request with specified start block and checkpoint block": {
			options: func() []ChaincodeEventsOption {
				checkpointer := new(InMemoryCheckpointer)
				checkpointer.CheckpointBlock(500)
				return []ChaincodeEventsOption{
					WithStartBlock(418),
					WithCheckpoint(checkpointer),
				}
			}(),
			expected: &gateway.ChaincodeEventsRequest{
				StartPosition: &orderer.SeekPosition{
					Type: &orderer.SeekPosition_Specified{
						Specified: &orderer.SeekSpecified{
							Number: 501,
						},
					},
				},
			},
		},
		"Sends valid request with specified start block and checkpoint transaction ID": {
			options: func() []ChaincodeEventsOption {
				checkpointer := new(InMemoryCheckpointer)
				checkpointer.CheckpointTransaction(500, "txn1")
				return []ChaincodeEventsOption{
					WithStartBlock(418),
					WithCheckpoint(checkpointer),
				}
			}(),
			expected: &gateway.ChaincodeEventsRequest{
				StartPosition: &orderer.SeekPosition{
					Type: &orderer.SeekPosition_Specified{
						Specified: &orderer.SeekSpecified{
							Number: 500,
						},
					},
				},
				AfterTransactionId: "txn1",
			},
		},
		"Sends valid request with no start block and fresh checkpointer": {
			options: []ChaincodeEventsOption{
				WithCheckpoint(new(InMemoryCheckpointer)),
			},
			expected: &gateway.ChaincodeEventsRequest{
				StartPosition: &orderer.SeekPosition{
					Type: &orderer.SeekPosition_NextCommit{
						NextCommit: &orderer.SeekNextCommit{},
					},
				},
			},
		},
		"Sends valid request with no start block and checkpoint transaction ID": {
			options: func() []ChaincodeEventsOption {
				checkpointer := new(InMemoryCheckpointer)
				checkpointer.CheckpointTransaction(500, "txn1")
				return []ChaincodeEventsOption{
					WithCheckpoint(checkpointer),
				}
			}(),
			expected: &gateway.ChaincodeEventsRequest{
				StartPosition: &orderer.SeekPosition{
					Type: &orderer.SeekPosition_Specified{
						Specified: &orderer.SeekSpecified{
							Number: 500,
						},
					},
				},
				AfterTransactionId: "txn1",
			},
		},
		"Sends valid request with with start block and checkpoint chaincode event": {
			options: func() []ChaincodeEventsOption {
				checkpointer := new(InMemoryCheckpointer)
				event := &ChaincodeEvent{
					BlockNumber:   1,
					TransactionID: "TRANSACTION_1",
				}
				checkpointer.CheckpointChaincodeEvent(event)
				return []ChaincodeEventsOption{
					WithCheckpoint(checkpointer),
				}
			}(),
			expected: &gateway.ChaincodeEventsRequest{
				StartPosition: &orderer.SeekPosition{
					Type: &orderer.SeekPosition_Specified{
						Specified: &orderer.SeekSpecified{
							Number: 1,
						},
					},
				},
				AfterTransactionId: "TRANSACTION_1",
			},
		},
	} {
		t.Run(testName, func(t *testing.T) {
			mockConnection := NewMockClientConnInterface(t)
			mockStream := NewMockClientStream(t)
			ExpectChaincodeEvents(mockConnection, WithNewStreamResult(mockStream))

			messages := make(chan *gateway.SignedChaincodeEventsRequest, 1)
			ExpectSendMsg(mockStream, CaptureSendMsg(messages))
			mockStream.EXPECT().CloseSend().Return(nil)
			ExpectRecvMsg(mockStream).Maybe().Return(io.EOF)

			ctx, cancel := context.WithCancel(context.Background())
			defer cancel()

			network := AssertNewTestNetwork(t, "NETWORK", WithClientConnection(mockConnection))
			_, err := network.ChaincodeEvents(ctx, "CHAINCODE", testCase.options...)
			require.NoError(t, err)

			creator, err := network.signingID.Creator()
			require.NoError(t, err)

			expected := &gateway.ChaincodeEventsRequest{
				ChannelId:   "NETWORK",
				ChaincodeId: "CHAINCODE",
				Identity:    creator,
			}
			proto.Merge(expected, testCase.expected)
			actual := &gateway.ChaincodeEventsRequest{}
			AssertUnmarshal(t, (<-messages).Request, actual)
			AssertProtoEqual(t, expected, actual)
		})
	}

	t.Run("Closes event channel on receive error", func(t *testing.T) {
		mockConnection := NewMockClientConnInterface(t)
		mockStream := NewMockClientStream(t)
		ExpectChaincodeEvents(mockConnection, WithNewStreamResult(mockStream))

		ExpectSendMsg(mockStream)
		mockStream.EXPECT().CloseSend().Return(nil)
		ExpectRecvMsg(mockStream).Maybe().Return(errors.New("fake"))

		ctx, cancel := context.WithCancel(context.Background())
		defer cancel()

		network := AssertNewTestNetwork(t, "NETWORK", WithClientConnection(mockConnection))

		receive, err := network.ChaincodeEvents(ctx, "CHAINCODE")
		require.NoError(t, err)

		actual, ok := <-receive

		require.False(t, ok, "Expected event listening to be cancelled, got %v", actual)
	})

	t.Run("Receives events", func(t *testing.T) {
		expected := []*ChaincodeEvent{
			{
				BlockNumber:   1,
				ChaincodeName: "CHAINCODE",
				EventName:     "EVENT_1",
				Payload:       []byte("PAYLOAD_1"),
				TransactionID: "TRANSACTION_ID_1",
			},
			{
				BlockNumber:   1,
				ChaincodeName: "CHAINCODE",
				EventName:     "EVENT_2",
				Payload:       []byte("PAYLOAD_2"),
				TransactionID: "TRANSACTION_ID_2",
			},
			{
				BlockNumber:   2,
				ChaincodeName: "CHAINCODE",
				EventName:     "EVENT_3",
				Payload:       []byte("PAYLOAD_3"),
				TransactionID: "TRANSACTION_ID_3",
			},
		}

		mockConnection := NewMockClientConnInterface(t)
		mockStream := NewMockClientStream(t)
		options := make(chan []grpc.CallOption, 1)
		ExpectChaincodeEvents(mockConnection, CaptureNewStreamOptions(options), WithNewStreamResult(mockStream))

		messages := make(chan *gateway.SignedChaincodeEventsRequest, 1)
		ExpectSendMsg(mockStream, CaptureSendMsg(messages))
		mockStream.EXPECT().CloseSend().Return(nil)
		ExpectRecvMsg(mockStream, WithRecvMsgs(
			newChaincodeEventsResponse(expected[0:2]),
			newChaincodeEventsResponse(expected[2:]),
		))

		ctx, cancel := context.WithCancel(context.Background())
		defer cancel()

		network := AssertNewTestNetwork(t, "NETWORK", WithClientConnection(mockConnection))

		receive, err := network.ChaincodeEvents(ctx, "CHAINCODE")
		require.NoError(t, err)

		for _, event := range expected {
			actual := <-receive
			require.Equal(t, event, actual)
		}
	})

	t.Run("Uses specified gRPC call options", func(t *testing.T) {
		expected := grpc.WaitForReady(true)

		mockConnection := NewMockClientConnInterface(t)
		mockStream := NewMockClientStream(t)
		options := make(chan []grpc.CallOption, 1)
		ExpectChaincodeEvents(mockConnection, CaptureNewStreamOptions(options), WithNewStreamResult(mockStream))

		messages := make(chan *gateway.SignedChaincodeEventsRequest, 1)
		ExpectSendMsg(mockStream, CaptureSendMsg(messages))
		mockStream.EXPECT().CloseSend().Return(nil)
		ExpectRecvMsg(mockStream).Maybe().Return(io.EOF)

		ctx, cancel := context.WithCancel(context.Background())
		defer cancel()

		network := AssertNewTestNetwork(t, "NETWORK", WithClientConnection(mockConnection))

		request, err := network.NewChaincodeEventsRequest("CHAINCODE")
		require.NoError(t, err, "NewChaincodeEventsRequest")

		_, err = request.Events(ctx, expected)
		require.NoError(t, err, "Events")

		require.Contains(t, <-options, expected, "CallOptions")
	})
}
